adapters/Vungle/VungleAdapter/Bidding/GADMediationVungleNativeAd.m

```diff
@@ -0,0 +1,248 @@
+                                     completionHandler:(nonnull GADMediationNativeLoadCompletionHandler)completionHandler {
+  self = [super init];
+  if (self) {
+    _adConfiguration = adConfiguration;
+
+    // Store the ad config and completion handler for later use.
+    __block atomic_flag adLoadHandlerCalled = ATOMIC_FLAG_INIT;
+    __block GADMediationNativeLoadCompletionHandler originalCompletionHandler = [completionHandler copy];
+
+    // Ensure the original completion handler is only called once, and is deallocated once called.
+    _adLoadCompletionHandler = ^id<GADMediationNativeAdEventDelegate>(_Nullable id<GADMediationNativeAd> ad, NSError *_Nullable error) {
+      if (atomic_flag_test_and_set(&adLoadHandlerCalled)) {
+        return nil;
+      }
+      id<GADMediationNativeAdEventDelegate> delegate = nil;
+      if (originalCompletionHandler) {
+        delegate = originalCompletionHandler(ad, error);
+      }
+      originalCompletionHandler = nil;
+      return delegate;
+    };
+  }
+  return self;
+}
+
+- (void)requestAd {
+  self.desiredPlacement = [GADMAdapterVungleUtils findPlacement:_adConfiguration.credentials.settings
+                                                  networkExtras:_adConfiguration.extras];
+  if (!self.desiredPlacement) {
+    NSError *error = GADMAdapterVungleErrorWithCodeAndDescription(GADMAdapterVungleErrorInvalidServerParameters, @"Placement ID not specified.");
+    _adLoadCompletionHandler(nil, error);
+    return;
+  }
+    
+  if (![[VungleSDK sharedSDK] isInitialized]) {
+    NSString *appID = [GADMAdapterVungleUtils findAppID:_adConfiguration.credentials.settings];
+    [[GADMAdapterVungleRouter sharedInstance] initWithAppId:appID delegate:self];
+    return;
+  }
+    
+  [self loadAd];
+}
+
+- (void)loadAd {
+  _nativeAd = [[VungleNativeAd alloc] initWithPlacementID:self.desiredPlacement];
+  _nativeAd.delegate = self;
+  VungleAdNetworkExtras *networkExtras = [_adConfiguration extras];
+  if (networkExtras.nativeAdOptionPosition != 0) {
+    _nativeAd.adOptionsPosition = networkExtras.nativeAdOptionPosition;
+  }
+  [_nativeAd loadAd];
+}
+
+#pragma mark - GADMediatedUnifiedNativeAd
+
+- (nullable NSString *)headline {
+    return _nativeAd.title;
+}
+
+- (nullable NSArray<GADNativeAdImage *> *)images {
+  return nil;
+}
+
+- (nullable NSString *)body {
+    return _nativeAd.bodyText;
+}
+
+- (nullable GADNativeAdImage *)icon {
+    return [[GADNativeAdImage alloc] initWithImage:_nativeAd.iconImage];
+}
+
+- (nullable NSString *)callToAction {
+    return _nativeAd.callToAction;
+}
+
+- (nullable NSDecimalNumber *)starRating {
+  return [[NSDecimalNumber alloc] initWithDouble:_nativeAd.adStarRating];
+}
+
+- (nullable NSString *)store {
+  return nil;
+}
+
+- (nullable NSString *)price {
+  return nil;
+}
+
+- (nullable NSString *)advertiser {
+  return _nativeAd.sponsoredText;
+}
+
+- (nullable NSDictionary<NSString *, id> *)extraAssets {
+  return nil;
```

Comment:
Looking at:
https://support.vungle.com/hc/en-us/articles/4416477036955#display-a-native-ad-0-5

There is "Sponsored By" text, does the adapter want to handle that as well?

Also see [`extraAssets`](https://developers.google.com/admob/ios/api/reference/Protocols/GADMediatedUnifiedNativeAd?hl=en#extraassets).

Reply:
Isn't that the `advertiser` property above?

Reply:
That's right, missed that one.